In [9]:
# Imports
import pandas as pd

### WIP

Ideas:
    1.add introduction
    2.add methods (?)
    3.provide brief introduction to the data
    4.


### WIP


### Context
"Predict behavor to retain customers. You can analyze all relevant customer data and develop focused customer retention programs."

### Content
Each row represents a customer, each column contains customer’s attributes described on the column Metadata.

The data set includes information about:

Customers who left within the last month – the column is called Churn
Services that each customer has signed up for – phone, multiple lines, internet, online security, online backup, device protection, tech support, and streaming TV and movies
Customer account information – how long they’ve been a customer, contract, payment method, paperless billing, monthly charges, and total charges
Demographic info about customers – gender, age range, and if they have partners and dependents

In [7]:
# Load & print initial info
df = pd.read_csv('data/WA_Fn-UseC_-Telco-Customer-Churn.csv')
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            7043 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   7043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       7043 non-null   str    
 14  StreamingMovies   7043 non-null   str    
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   str    
 17  Paymen

In [8]:
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


### WIP add short data description

It's possible to see our dataset has 21 columns/features and 7,043 observations. Most of the features are made of string data type, including 'TotalCharges', which appear to be made of numbers and not letters, and we have 3 columns made of numbers, either as integers or float64. It is possible to see that we have a substantial number of features made of categorical data, which will increase the number of columns after treatment for use with statistical models.



In [13]:
len(set(df['customerID']))



7043

In [ ]:
# Check string data
string_cols = [
    'customerID',
    'gender',
    'Partner',
    'Dependents',
    'PhoneService',
    'MultipleLines',
    'InternetService',
    'OnlineSecurity',
    'OnlineBackup',
    'DeviceProtection',
    'TechSupport',
    'StreamingTV',
    'StreamingMovies',
    'Contract',
    'PaperlessBilling',
    'PaymentMethod',
    'TotalCharges',
    'Churn' 
]

num_cols = [
    'SeniorCitizen',
    'tenure',
    'MonthlyCharges',
]

for col in string_cols:
    print(f"{col}\n{len(col)*'-'}")
    print(set(df[col]), '\n')

customerID
----------
{'2368-GAKKQ', '6901-GOGZG', '7967-HYCDE', '2750-BJLSB', '0388-EOPEX', '4226-KKDON', '5277-ZLOOR', '4188-PCPIG', '0708-LGSMF', '0238-WHBIQ', '5207-PLSTK', '8029-XYPWT', '2223-KAGMX', '1905-OEILC', '6507-ZJSUR', '8992-VONJD', '3577-AMVUX', '0679-IDSTG', '8402-EIVQS', '2761-XECQW', '4615-PIVVU', '8838-GPHZP', '2155-AMQRX', '9692-TUSXH', '1170-SASML', '0519-XUZJU', '2362-IBOOY', '0786-VSSUD', '3498-LZGQZ', '8091-TTVAX', '9554-DFKIC', '5510-BOIUJ', '3458-IDMFK', '0197-PNKNK', '3853-LYGAM', '6481-OGDOO', '9378-FXTIZ', '3156-QLHBO', '3863-IUBJR', '0635-WKOLD', '7748-UMTRK', '2085-BOJKI', '3813-DHBBB', '9938-PRCVK', '6705-LNMDD', '2176-LVPNX', '9031-ZVQPT', '9027-TMATR', '7839-QRKXN', '8008-HAWED', '8566-YPRGL', '9220-CXRSC', '4547-FZJWE', '5136-KCKGI', '4830-FAXFM', '0118-JPNOY', '2027-FECZV', '5787-KXGIY', '7820-ZYGNY', '9495-REDIY', '6434-TTGJP', '7562-UXTPG', '2043-WVTQJ', '9822-WMWVG', '2303-PJYHN', '9610-WCESF', '4360-PNRQB', '6685-XSHHU', '8605-ITULD', '4176-FXYBO

In [18]:
for col in num_cols:
    print(f"{col}\n{len(col)*'-'}")
    print(set(df[col]), '\n')

SeniorCitizen
-------------
{0, 1} 

tenure
------
{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72} 

MonthlyCharges
--------------
{18.95, 19.8, 20.65, 20.15, 20.2, 20.75, 24.95, 20.4, 24.25, 25.9, 19.95, 29.75, 29.85, 30.2, 24.3, 33.75, 30.15, 31.05, 35.45, 29.95, 30.4, 39.65, 35.9, 40.2, 42.3, 42.1, 44.35, 45.25, 45.55, 45.65, 45.3, 49.95, 49.55, 49.25, 49.05, 53.85, 54.4, 55.2, 56.95, 56.15, 55.3, 59.9, 59.6, 55.65, 54.65, 62.15, 64.7, 64.15, 66.15, 64.5, 66.85, 69.7, 70.7, 69.2, 69.5, 70.45, 74.8, 74.85, 76.2, 75.3, 78.9, 79.85, 80.65, 79.75, 79.2, 74.75, 84.5, 79.35, 82.05, 84.0, 85.2, 89.1, 90.05, 90.25, 89.9, 91.65, 94.4, 95.5, 96.35, 97.85, 95.45, 99.65, 100.35, 99.35, 102.45, 102.7, 103.8, 103.4, 104.15, 104.4, 104.6, 105.35, 108.0, 109.2, 103.7,